File: C++/addTwoNum/addTwoNum.cpp
#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
  public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        int x = 0, y = 0, carry = 0, sum = 0;
        ListNode *h = NULL, **t = &h;

        while (l1 != NULL || l2 != NULL)
        {
            x = getValueAndMoveNext(l1);
            y = getValueAndMoveNext(l2);

            sum = carry + x + y;

            ListNode *node = new ListNode(sum % 10);
            *t = node;
            t = (&node->next);

            carry = sum / 10;
        }

        if (carry > 0)
        {
            ListNode *node = new ListNode(carry % 10);
            *t = node;
        }

        return h;
    }

    void output(ListNode *head)
    {
        while (head != NULL)
        {
            cout << head->val << endl;
            head = head->next;
        }
    }

    void free(ListNode *head)
    {
        cout << "不知是否需要手动释放链表的内存？这里释放了" << endl;
        ListNode *next;
        while (head != NULL)
        {
            next = head->next;
            delete head;
            head = next;
        }
    }

  private:
    int getValueAndMoveNext(ListNode *&l)
    {
        int x = 0;
        if (l != NULL)
        {
            x = l->val;
            l = l->next;
        }
        return x;
    }
};

int main()
{
    // 第一个链表 l1
    ListNode *l1head = new ListNode(2);
    ListNode *l1second = new ListNode(4);
    ListNode *l1third = new ListNode(3);

    l1head->next = l1second;
    l1second->next = l1third;

    // 第二个链表 l2
    ListNode *l2head = new ListNode(5);
    ListNode *l2second = new ListNode(6);
    ListNode *l2third = new ListNode(4);

    l2head->next = l2second;
    l2second->next = l2third;

    Solution solution;
    ListNode *result = solution.addTwoNumbers(l1head, l2head);
    solution.output(result);
    solution.free(l1head);
    solution.free(l2head);
    solution.free(result);
    return 0;
}

File: Java/addTwoNum/AddTwoNum.java
/**********************************************************************************
 * 
 * You are given two linked lists representing two non-negative numbers. The
 * digits are stored in reverse order and each of their nodes contain a single
 * digit. Add the two numbers and return it as a linked list.
 * 
 * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4) Output: 7 -> 0 -> 8
 * 
 **********************************************************************************/
/**
 * 
 * author : cjc
 * 2018 12 05 16:19
 * 
 */
class ListNode {
    private int val;
    private ListNode next;

    public ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    public void output() {
        System.out.println("this.val = " + this.val);
    }

    public ListNode getNext() {
        return this.next;
    }

    public void setNext(ListNode next) {
         this.next = next;
    }

    public int getVal() {
        return this.val;
    }
}

public class AddTwoNum {
    public ListNode addTwoNum(ListNode l1, ListNode l2) {
        ListNode h = null;
        ListNode forward = null;
        ListNode current = null;
        int sum=0, carry=0, x=0,y=0;

        while(l1 != null && l2 != null) {
            x = l1.getVal();
            y = l2.getVal();

            sum = carry + x + y;

           current = new ListNode(sum%10, null);
            if(h == null) {
                h = forward = current;
            } else {
                forward.setNext(current);
                forward = current;
            }
            carry = sum / 10;
            l1 = l1.getNext();
            l2 = l2.getNext();
        }
        return h;
    }

    public static void main(String[] args) {
        ListNode l1head = new ListNode(2, null); 
        ListNode l1sec = new ListNode(4, null);
        ListNode l1third = new ListNode(3, null);

        l1head.setNext(l1sec);
        l1sec.setNext(l1third);

        ListNode l2head = new ListNode(5, null);
        ListNode l2sec = new ListNode(6, null);
        ListNode l2third = new ListNode(4, null);

        l2head.setNext(l2sec);
        l2sec.setNext(l2third);

        AddTwoNum addTwoNum = new AddTwoNum();
        ListNode listNode =  addTwoNum.addTwoNum(l1head, l2head);
        addTwoNum.printlist(listNode);
        
    }

    public void printlist(ListNode head) {
        while(head != null) {
            head.output();
            head = head.getNext();
        }
    }
}